### ISSUE TITLE
APNG/animated GIF support 

### ISSUE THREAD

**Issue by NOVAinc** [pubpub-zz/ppInk#101]
Hi again! I have come up with a few more ideas that I think could be helpful. Once again, both of these things accommodate to my use-case specifically, but transparency in the window is something that anyone using the tool with OBS will be able to benefit from.

When using Game Capture source in OBS there is an option to allow for transparency, but ppink doesn't work with this capture method, as it currently works only with Window Capture, in Windows Graphis Capture mode. It would be great having the ability to use Game Capture so we could take advantage of translucency in strokes since otherwise we must use a color key for the currently black background.

As for APNG, I'm not sure what library is being used for ClipArts, but there is no current support for APNG, or WebP. Both these formats support lossless (and webp also lossy) animations with a much wider color gamut and better transparency support. APNG specifically even yields smaller sizes for the same animations as gifs. Having either APNG or webp support would be as close as having transparent video support for cliparts, which might be harder to implement unless using AVI or something like that. Here is a comparison of the formats in 5 different scenes.

http://littlesvr.ca/apng/gif_apng_webp1.html

**Comment by pubpub-zz** [comment]
Hi @NOVAinc,
About game capture, I did my best To get all features. there has been lots of exchange in #21. I don't think I have introduced regression about (except the window name. with the workaround found). I've not been able to find a solution on the web (eg. https://obsproject.com/forum/threads/capture-transparent-c-wpf-window.75380/ ) If you know a c# program that is working with this feature I can have a look at the code.

About APNG, This is an idea, however, keep in mind ppInk is updating screen based on a 15ms timer and which may not be very stable. However it worth to have a try 👍

**Comment by NOVAinc** [comment]
Oh, I see. So this means that a gif for instance would refresh around 60 times a second instead of following the gif's time stamps?

As for game capture, I don't know of any programs that one would have access to the code to take a look at, alas.

As for the recent change in the "Drawing in progress..." window, I realized that even when using "Window must match the exact title" in OBS the program doesn't seem to be able to find it if you close the toolbar, switch to a different scene, and then try to load it back up. What I've been doing instead is I've been collapsing the toolbar whenever I stream so that the expand arrow would still always show up and that way the Drawing in progress window never disappears. I do still have to configure the source manually every time I use it for the first time on a stream, while before this wasn't an issue. It isn't anything major, really, so I hadn't said anything but I guess it's probably a good idea just to let you know it is happening :)

**Comment by pubpub-zz** [comment]
For performane reasons, the ppink drawing window is no more destroyed but hidden. I will have to make more analysis with OBS. not for right now

**Comment by pubpub-zz** [comment]
@NOVAinc,
This an alpha that you should like : It introduces APNG compatibility 😎 
Keep in mind that apng increases cpu loading and that's why I've stop animation during drawing.
[ppInk64_apng16a.zip](https://github.com/pubpub-zz/ppInk/files/6672338/ppInk64_apng16a.zip)
If you have a few free-to-use  images to add for demo in ppInk delivery It would be great 😃

**Comment by NOVAinc** [comment]
I have made this special clipart to showcase the apgn capabilities. Increased color range and granular translucency as opposed to full opaque/full transparent on gifs! Thanks a lot for implementing this feature 😄 

![ppink-cliparts](https://user-images.githubusercontent.com/19545604/122837671-dbee8200-d2ca-11eb-934b-9a059ab4b202.png)

**Comment by pubpub-zz** [comment]
> -The first one is that it had a firefox.png picture linked in the clipart gallery but it didn't come packaged with the app so it throws an error on lunch.

Yes I forgot to remove ref to my test,will remove it for next beta/releases

> -I've seen artifacts using this same clipart I made in which it gets distorted and accelerates the animation in the middle of the circle. This might be a bug with the library seeing how the same file seems to work well in the browser here. Let me know if you need anything. I'll attach a zip with all the frames of the animation if it is helpful for debugging. Let me know if you need anything else.

image jump seem to be due to my index reset at the end of sequence. should be good now. About  'spinning speed not sure where it is coming from performances or decoding. What I've notice is that the rate is very high. may be the issue, can you try it with a frame rate of 50ms less.

> -I've also tried to use a gif picture found [here](https://github.com/Explosion-Scratch) and it doesn't seem to work as only the first frame is displayed in a static manner. I'll attach a video with both these issues.

the library I'm using ( https://github.com/ImoutoChan/ApngWpfPlayer ) is only processing apng. Not sure yet I will implement gif also : 
     * I will have to implement a inherited object to interface both ;
     * animated gif has very limited implementation of transparency
     * many converters are available ( personnaly I've tryied RealWorld Paint)

I converted the explosion you've refered and results looks good for me:
![explosion](https://user-images.githubusercontent.com/4083478/123539306-568b2780-d739-11eb-8955-eced8e804ba1.png)

> It'd also be good to have a toggle for apng animation during drawing too since on some PCs CPU usage might not be an issue. With the clip running on my PC CPU goes from 3% to 5% for ppink so it'd be handy being able to just let it run

No worries. ppInk only do animation on APNG. if it is a standard PNG, legacy drawing is implemented. For me it is just the user to load animation only on adequate plaftforms.

**Comment by pubpub-zz** [comment]
@NOVAinc 
forgot : latest version:
[ppInk64_apng27a.zip](https://github.com/pubpub-zz/ppInk/files/6721339/ppInk64_apng27a.zip)

**Comment by NOVAinc** [comment]
My bad, I meant to report gif as being broken on this release, I didn't know they weren't supported. I assumed they were because before APNG support was added when you went to import cliparts the gif format showed in the list of images you could import, but I hadn't tested them so far. I will test the new release later today and I'll post my results.

**Comment by NOVAinc** [comment]
I tested it a bit more now. I'm not really sure how to make the frame delay 50ms, since ezgif which I used to make the clipart one didn't have a ms field. Instead it had a generic "delay" field which I used to create 5 variants of the animation with delays of 1, 2, 16, 33 and 60 (I thought these could have been ms values but 1 delay isn't 1/1000th of a second so it might be Hz... however, at 60 it doesn't do 1fps so I have no idea what the delay is based on)

I tried to import the explosion png file you shared above as well. With the explosion there is no deformation issue, but with the clipart rotating piece it still happens. I did notice, however, that the explosion seems to be "jitters" a bit and it plays out faster than in the browser. As you can see with ppink's explosion (left) it starts playing at the same time as the browser version, but eventually gets ahead. To the right there are the 5 versions I mentioned of the spinning clipart.

https://user-images.githubusercontent.com/19545604/123561285-c1f6e880-d77d-11eb-8159-ee6768b7682c.mp4

**Comment by pubpub-zz** [comment]
I finally found a solution for animated gifs! 
[ppInk64_apng 28a.zip](https://github.com/pubpub-zz/ppInk/files/6729439/ppInk64_apng.28a.zip)

I've that way been able to test the rotating ppInk with animated gif : I've then noticed the rendering between the apng and gif are different. 
![image](https://user-images.githubusercontent.com/4083478/123709492-73c11280-d86d-11eb-93cc-216a79f78452.png)

It looks like the decoding library for apng is buggy 🤔 
I will investigate more

be aware also that the restore stroke is crashing for me with animation...

**Comment by NOVAinc** [comment]
I was going to suggest maybe it was stretching the narrower frames to occupy the entire PNG canvas, but I thought that couldn't be it because the explosion doesn't distort when it's not occupying the entire canvas.

Good job! I will test the new version for the gif functionality. It might be lighter weight than APNG but of course it is going to yield larger file sizes and a reduced color palette/transparency support. It is easier to find gif assets than APNG for the general public as well, though, so it seems like a great addition in general!

**Comment by pubpub-zz** [comment]
@NOVAinc 
found and solved !
[ppInk64_apng29a.zip](https://github.com/pubpub-zz/ppInk/files/6736157/ppInk64_apng29a.zip)
I've also fixed the load stroke
Performances looks good on my side. OK for you ?
Currently, I've implemented infinite loop ; I would not like to complexify the interface, therefore I consider that  a limited number of play is not required. your opinion?

**Comment by NOVAinc** [comment]
It works! and it's amazing! I don't think it's necessary to limit the playback for infinite loops, however it is possible to include a limit to the amount of times an animated png can cycle through. For example, this version below is supposed to spin only 5 times, but with ppink it seems to spin forever. I personally don't mind because the loops I'll personally be using are inifinite, but I thought I should report my findings.

Here is the example.

![ezgif com-gif-maker](https://user-images.githubusercontent.com/19545604/123908727-5021c900-d94e-11eb-9f96-aa44dd33624e.png)

**Comment by pubpub-zz** [comment]
I thought a little more. I do not think everybody will have the editors to change the internal values :
this is my proposal  : if the file name finishes with a value in square brackets, we will use to define how long (in decimal seconds) ppInk will animate it. if the value is negative, the image will be deleted - instead of just stopping the animation. without info it will be for ever.
if clicks are performed before clicking, it will reset counters (for all currently animated images).
Your opinion ?

**Comment by NOVAinc** [comment]
That is a really good workaround. Do you think it'd be possible to use maybe a different notation for amount of times the loop takes place? like [x1] at the end and the x would indicate when it's an indicator for a number of cycles instead of seconds. Your proposal seems sensible and with some tinkering anyone might be able to rename their files so the time in seconds matches the amount of cycles they want. It's a good idea-

**Comment by pubpub-zz** [comment]
Seems feasable, but to ease decoding, I will put the x at the end.
ppInkClipart.png -> animated for ever
ppInkClipart[5.2].png -> animated for 5.2 sec and then animation stops
ppInkClipart[-3.1].png -> animated for 3.1 sec and then disappear
ppInkClipart[3x].png -> animated for 3 cycles and then animation stops
ppInkClipart[-2x].png -> animated for 2 cycles and then disappear

**Comment by pubpub-zz** [comment]
By the way, what license applies to the rotating apng file you've provided?

**Comment by NOVAinc** [comment]
That sounds awesome! The rotating apng I made is for you. Do with it what you please.

**Comment by pubpub-zz** [comment]
Hi @NOVAinc,
Here it is the full implementation for APNG/Animated Gif.
[ppInk64_apng_alt_dash_07a.zip](https://github.com/pubpub-zz/ppInk/files/6756735/ppInk64_apng_alt_dash_07a.zip)

I will fork this issue to take appart the transparency
Also can you give me your opinion about merging es-es and es-latam proposed in https://github.com/pubpub-zz/ppInk/issues/17#issuecomment-860182551
Thanks

**Comment by NOVAinc** [comment]
I have done some testing and Y, Yx, -Y and -Yx seem to work really well. The only scenarios in which I haven't seen it work as expected is when using fractions for repetition amount. Such as 2.5x or -2.5x. I don't think this is necessary and I just did it for testing purposes, but it might be a good idea to mention in the documentation that fractional amounts aren't supported for repetition times. Really good job! And to think I suggested APNG support _thinking_ gifs were already supported. Now both are. Amazing stuff!

**Comment by pubpub-zz** [comment]
@NOVAinc 
I originally was not sure about not int loops, but I found how to implement it. try this:
[ppInk64_apng04a.zip](https://github.com/pubpub-zz/ppInk/files/6759542/ppInk64_apng04a.zip)

PS: can you tell me if you are experiencing same type of problems as the one reported in #97
Thanks

**Comment by NOVAinc** [comment]
The new version doesn't seem to take fractional loops from my attempt. Instead it seems ignore the point so if it is 1.5x it'll loop 15 times. I've also noticed it seems to render one too many frames per loop as you can see 1x and 5x in the video below stop on a different frame.


https://user-images.githubusercontent.com/19545604/124529240-79c07180-dde0-11eb-8137-7b860ac39dcb.mp4

**Comment by pubpub-zz** [comment]
a) I forgot to parse with invariant culture : should fix the 15 times. pls confirm
b) I made some mistakes in my counting of frames : should fix the offset in the number of frames at the end pls confirm
c) Be aware that this version is including some to improve hick ups (see https://github.com/pubpub-zz/ppInk/issues/97#issuecomment-874348042) do not hesitate to report problems

[ppInk64_apng06a.zip](https://github.com/pubpub-zz/ppInk/files/6769724/ppInk64_apng06a.zip)

**Comment by NOVAinc** [comment]
Nice! The new version seems to work with positive loop counts. I tested 0.5x and it stops halfway, however it doesn't make it disappear if it is a negative cycle count. In this example -0.5x doesn't disappear after half a cycle. But it also doesn't go up to 5 cycles which it did before, so it's better still.

Also, the frame is now consistent for an arbitrary number of cycles, so both 1x and 5x are stopping on the same frame!

https://user-images.githubusercontent.com/19545604/124649142-7f649880-de6e-11eb-825f-e7523558d2ed.mp4

**Comment by pubpub-zz** [comment]
Arg!!!!
When I've allowed decimal, the negative number has been removed....
Fixed now:
[ppInk64_apng_perfo06b.zip](https://github.com/pubpub-zz/ppInk/files/6772517/ppInk64_apng_perfo06b.zip)
Also this version includes some complementary changes on the drawing for the hickups. tell me if it loos still good for drawing

**Comment by NOVAinc** [comment]
It works now! The only issue I've found this time is they all seem to snap back to the first frame when moving one
https://user-images.githubusercontent.com/19545604/124668577-759b5f00-de87-11eb-8303-ad1b44858d74.mp4

**Comment by pubpub-zz** [comment]
image 1 was displayed during moving. I've normally fixed it. capture in transparent mode (stroke only) was also forgotten. fixed also
About other drawings have you seen any odd thing/regressions with my latest mods?

If not I Thing we have a Release candidate. You will be able to make soon your demo video 😊
[ppInk64_apng06c.zip](https://github.com/pubpub-zz/ppInk/files/6773062/ppInk64_apng06c.zip)

**Comment by NOVAinc** [comment]
This new version is working really well for me with no regressions. I have had some small issues here and there when using the tool, such as the cycles stopping one frame too late some times, and then some other issues raising errors from time to time BUT I have been unable to reproduce these, so I can't really be of much help. As I encounter issues in the future I will try to find ways to reproduce them so I can help you more. For the time being they're just sparse and far between so it's not a problem I'd say. The issues I encountered still don't prevent users from using the tool as it's only crashed on me once when using the stroke style hotkey while having a different tool selected but as stated I couldn't reproduce it anymore. If I get any more issues I'll make sure to report them!

**Comment by pubpub-zz** [comment]
Sound good. I will then release this version tonight. about the issues you've met, even if you don't have logs/details, can you at least describe them inhere,just for loging

**Comment by NOVAinc** [comment]
Sure thing.

- The first time I opened the latest version you shared above, I got an error sign (UI Exception if I'm not mistaken) but after clicking Ok ppink started as normal. Restarting ppink wouldn't bring the issue up anymore.

- During a stream the other day I was drawing some strokes with a dashed style (I don't recall exactly which one) and then I tried to add a clipart. After doing this I noticed the stroke style stayed changed for the pencil I was using beforehand so I tried to use the stroke style hotkey to cycle it back to standard stroke and ppink crashed. I couldn't replicate this one either.

- Yesterday while testing the lastest version, I inserted the clipart in 5x, .5x, -.5x and 1x in rapid succession and the 5x version ended the loop one frame off compared to 1x. I couldn't replicate this one when I tried again as the frames would be the same for both.

**Comment by pubpub-zz** [comment]
>     * The first time I opened the latest version you shared above, I got an error sign (UI Exception if I'm not mistaken) but after clicking Ok ppink started as normal. Restarting ppink wouldn't bring the issue up anymore.

there should be a crash.txt file in ppInk folder, can you send it pls

**Comment by NOVAinc** [comment]
There we go!

[crash.txt](https://github.com/pubpub-zz/ppInk/files/6779741/crash.txt)

**Comment by pubpub-zz** [comment]
@NOVAinc 

>     * The first time I opened the latest version you shared above, I got an error sign (UI Exception if I'm not mistaken) but after clicking Ok ppink started as normal. Restarting ppink wouldn't bring the issue up anymore.

Sometimes the startup is slowed (why? don't know). In those cases using the shortcut or the trayicon was raising the exception. I've reordered init to fix that

>     * During a stream the other day I was drawing some strokes with a dashed style (I don't recall exactly which one) and then I tried to add a clipart. After doing this I noticed the stroke style stayed changed for the pencil I was using beforehand so I tried to use the stroke style hotkey to cycle it back to standard stroke and ppink crashed (I imagine because I had clipart selected as opposed to the pen I tried to alter). I couldn't replicate this one either.

Can not reproduce. Do you have the crash.txt ?

>     * Yesterday while testing the lastest version, I inserted the clipart in 5x, .5x, -.5x and 1x in rapid succession and the 5x version ended the loop one frame off compared to 1x. I couldn't replicate this one when I tried again as the frames would be the same for both.

this could be possible.the frame change is performed on a internal timer that could be affected by drawing cpu load and in those case you may not stop on frame 1.

> 
>     * With the latest versions, I was having trouble re-adding cliparts to the gallery that I had accidentally deleted before. I.e adding 1x and then deleting it, then trying to add it again would result in no addition.

Reproduced and fixed.

I've seen that you had an other crash that I guess to be due to quick and fast clicks : it is normally fixed also

here you are the new beta version
[ppInk64_varFix07a.zip](https://github.com/pubpub-zz/ppInk/files/6780232/ppInk64_varFix07a.zip)

**Comment by NOVAinc** [comment]
Here is the crash log for the stream crash! Hopefully it has more information regarding what the exact procedure was to help reproduce

[crash.txt](https://github.com/pubpub-zz/ppInk/files/6780244/crash.txt)

I will be testing the new beta version!

BTW: do you have a milestone for when a stable version is released? I've seen in the release page there are a lot of release candidates but I don't think I've seen a formal release yet. This is just me being curious as I understand it is always work in progress in special when adding so much functionality so fast!

**Comment by pubpub-zz** [comment]
Can not correlate the second crash from crash.txt with the existing repo code... Will have to wait new occurrence...

About versions I consider each version in Release as official. In some cases (eg translation,minor fixes, enhancement) or urgent fixes I issue a new version without increasing the version (ex 1.4alpha1 to 1.4alpha1a) ,the qualification alpha,beta,RC is an indication for me of the reliability of this version (It most of the time mean I had lots of test reports on the beta release I issued)

next version will be **1.7** _RC1_ 😊

**Comment by NOVAinc** [comment]
Thanks for the response. I have tested the version you shared above and so far I've seen no regressions. The issue re-adding cliparts is now resolved indeed, and I've been unable to make it crash when tweaking stroke styles!

PS/ Do you have any special plans in mind for version 2.0?

**Comment by pubpub-zz** [comment]
> PS/ Do you have any special plans in mind for version 2.0?

No.  In my mind v2.0 would mean a major break that could occur because of environment change : 
              maybe to switch to a higher .net framework ? but I do not see any functional advantage to switch to a higher version 
              Change to OpenGL or DirectX (for OBS) could be a reason. but I'm not ready yet for that
any idea of a functional change that would required such a break?

**Comment by NOVAinc** [comment]
I do have some functionality ideas as well as some suggestions for a potential version 2.0! Would you like me to open a new issue or just post them here?

The reason I was keen on having transparency support for OBS was because I'd be able to use black strokes or black elements which now don't appear since I have to color-key black when using ppink as an independent source as opposed to Display Capture. I have, however, thought of a way to work around this issue using ppink's existing codebase.

If you add ppink as an independent source, and then you apply a LUT to it, you can make it so that OBS displays certain colors certain ways. This way, you could create a LUT to have OBS interpret a specific hue as black, and then you could use that color for stroke or as part of your cliparts and OBS would change it to black. The downside to this is that you then wouldn't be able to display that color without it having appear black. I think, though, that it might be possible to have an LUT that darkens shades of gray. This way, pure black could be color keyed while a fairly dark gray could be used on its place and have OBS make that shade of gray darker so it appears black. Because it would use a LUT as opposed to standard hue/gamma/saturation/contrast colors, it wouldn't affect any of the other colors on the canvas.

**Comment by pubpub-zz** [comment]
For the moment what about using nearly black R=1 G=1 B=1?

**Comment by pubpub-zz** [comment]
@NOVAinc ,
I've already opened #113.
The solution is not tricky. Obs inputs are not so numerous. I've already tried to implement a solution to periodically refresh an image that can be displayed in Obs. Obs does refresh the image, but refresh rate was not good. The only solution I've found is to use OpenL or DirectX, but for the moment I don't know how to do it (as said in #113), I'm looking for help....
